packages/react-pages/src/node/dynamic-modules/PagesData.ts

```diff
@@ -0,0 +1,289 @@
+      getStaticData: (
+        pageId: string
+      ): {
+        [key: string]: any
+      } =>
+        this.createMutableProxy(pageId, 'static', sourceFile, scheduleUpdate),
+    }
+  }
+
+  /**
+   * If the page page does not come from local file (.e.g from remote database),
+   * then we can't track the source of page data.
+   */
+  createAPIForCustomSource(scheduleUpdate: ScheduleUpdate): HanlderAPI {
+    return {
+      getRuntimeData: (
+        pageId: string
+      ): {
+        [key: string]: string
+      } => this.createMutableProxy(pageId, 'runtime', null, scheduleUpdate),
+      getStaticData: (
+        pageId: string
+      ): {
+        [key: string]: any
+      } => this.createMutableProxy(pageId, 'static', null, scheduleUpdate),
+    }
+  }
+
+  private createMutableProxy(
+    pageId: string,
+    dataType: 'runtime' | 'static',
+    sourceFile: File | null,
+    scheduleUpdate: ScheduleUpdate
+  ): {
+    [key: string]: any
+  } {
+    const _this = this
+    const dataObj = this.getDataObj(pageId, dataType)
+    return new Proxy(dataObj, {
+      set(target, prop: string, value) {
+        if (value === undefined) {
+          // undefined means 'delete'
+          const existValue = dataObj[prop]
+          if (existValue) {
+            existValue.unBind(scheduleUpdate)
+          }
+          return true
+        }
+        new Association(
+          _this,
+          pageId,
+          dataType,
+          prop,
+          value,
+          sourceFile,
+          scheduleUpdate
+        )
+        return true
+      },
+      get(target, prop: string) {
+        const association = target[prop]
+        if (association === undefined) return undefined
+        return association.value
+      },
+    })
+  }
+
+  toPagesData(): PagesData {
+    const result: PagesData = {}
+    Object.entries(this.pages).forEach(([pageId, page]) => {
+      if (this.isEmptyPage(pageId)) return
+
+      const runtimeData = Object.fromEntries(
+        Object.entries(page.runtimeData).map(([key, value]) => [
+          key,
+          value.value,
+        ])
+      )
+      const staticData = Object.fromEntries(
+        Object.entries(page.staticData).map(([key, value]) => [
+          key,
+          value.value,
+        ])
+      )
+      result[pageId] = {
+        data: runtimeData,
+        staticData,
+      }
+    })
+    return result
+  }
+
+  isEmptyPage(pageId: string) {
+    const page = this.pages[pageId]
+    if (!page) return true
+    const { runtimeData, staticData } = page
+    return (
+      Object.keys(runtimeData).length === 0 &&
+      Object.keys(staticData).length === 0
+    )
+  }
+
+  private initPageIfNotExist(pageId: string) {
+    let page = this.pages[pageId]
+    if (!page) {
+      page = this.pages[pageId] = { runtimeData: {}, staticData: {} }
+    }
+    return page
+  }
+
+  getDataObj(pageId: string, type: 'runtime' | 'static') {
+    const page = this.initPageIfNotExist(pageId)
```

Comment:
Maybe simpler to just inline this method, since it's only used here?
```suggestion
    const page = this.pages[pageId] || (this.pages[pageId] = { runtimeData: {}, staticData: {} })
```

Reply:
I think the current way is more readable. It worth a couple more lines of code.